### PR TITLE
Analyze Guice program nodes through their annotated type

### PR DESCRIPTION
**Pull request by jaysonacosta**

### CODE REVIEW

File: src/main/java/org/checkerframework/checker/dependencyinjection/DependencyInjectionTransfer.java [hunk 1]
```diff
@@ -48,31 +47,68 @@ public TransferResult<CFValue, CFStore> visitMethodInvocation(
...
+
+      CFValue value =
+          diATF
+              .getStoreBefore(node)
+              .getValue(JavaExpression.fromNode(methodInvocationNode.getTarget().getReceiver()));
+
+      if (value != null && !value.getAnnotations().isEmpty()) {
+        AnnotationMirror anno = value.getAnnotations().first();
+        List<String> classNames =
+            AnnotationUtils.getElementValueArray(anno, diATF.bindValValueElement, String.class);
```
Comment (kelloggm) [inline src/main/java/org/checkerframework/checker/dependencyinjection/DependencyInjectionTransfer.java:63]: This line makes the strong assumption that the first annotation will definitely be `@Bind`. I think you should test that before calling `getElementValueArray`: it's probably true almost all the time (`@Bind({})` is the default), but there's no guarantee that e.g. the user hasn't provided a weird stub file or something like that. You should always test what kind of annotation you have before calling `getElementValue` or `getElementValueArray`.

File: src/main/java/org/checkerframework/checker/dependencyinjection/DependencyInjectionTransfer.java [hunk 2]
```diff
@@ -48,31 +47,68 @@ public TransferResult<CFValue, CFStore> visitMethodInvocation(
...
+        accumulateBindAnnotatedWith(node, result, classNames.get(0), givenName.getValue());
+      }
+    }
+
+    return result;
+  }
+
+  @Override
+  public TransferResult<CFValue, CFStore> visitAssignment(
+      AssignmentNode node, TransferInput<CFValue, CFStore> input) {
```
Comment (kelloggm) [inline src/main/java/org/checkerframework/checker/dependencyinjection/DependencyInjectionTransfer.java:77]: What's the purpose of this code? I don't think it should be necessary to handle dataflow through assignments, which should already be handled by `CFAbstractTransfer`. That is, all checkers automatically will transfer any information about the right-hand side of an assignment to the left-hand side: you don't need to write extra code to achieve that.

If this is doing something else, it needs a comment to explain its purpose.

File: src/main/java/org/checkerframework/checker/dependencyinjection/qual/Bind.java [hunk 3]
```diff
@@ -8,22 +8,22 @@
...
+ * Represents classes and names that have been passed as arguments to {@link
+ * com.google.inject.Binder#bind()}.
  */
 @Retention(RetentionPolicy.RUNTIME)
 @Target({ElementType.TYPE_USE, ElementType.TYPE_PARAMETER})
 @SubtypeOf({})
 @DefaultQualifierInHierarchy
 public @interface Bind {
+  /** The fully-qualified name of this annotation. */
+  static final String NAME = "org.checkerframework.checker.dependencyinjection.qual.Bind";
```
Comment (kelloggm) [inline src/main/java/org/checkerframework/checker/dependencyinjection/qual/Bind.java:20]: This is a good idea, and we should probably do this is a matter of course for all new annotations. I like it :)

File: src/main/java/org/checkerframework/checker/dependencyinjection/qual/Bind.java [hunk 4]
```diff
@@ -8,22 +8,22 @@
...
   /**
-   * Classes that have definitely been passed as an arugment to a {@code bind} call whose result is
-   * the expression whose type is annotated. The arguments are "fully qualified binary names"
-   * ({@link org.checkerframework.checker.signature.qual.FqBinaryName}): a primitive or <a
-   * href="https://docs.oracle.com/javase/specs/jls/se17/html/jls-13.html#jls-13.1">binary name</a>,
-   * possibly followed by some number of array brackets.
+   * Class names that have definitely been passed as an argument to a {@link
+   * com.google.inject.Binder#bind()} call.
```
Comment (kelloggm) [inline src/main/java/org/checkerframework/checker/dependencyinjection/qual/Bind.java:23]: Why did you remove the note about the format of the argument? I think it is important to document it.

Also, this documentation should note that you expect that this will only ever have zero or one elements in the array, but that you used an array for compatibility with `AccumulationAnnotatedTypeFactory`.

**Review by kelloggm — CHANGES_REQUESTED** [review]
I'm not exactly sure what's causing the fluent code not to work, but I noticed a couple of issues. Happy to take another look later if fixing those doesn't resolve the problem.

File: src/main/java/org/checkerframework/checker/dependencyinjection/DependencyInjectionTransfer.java (hunk 2 above)
Comment (jaysonacosta) [reply src/main/java/org/checkerframework/checker/dependencyinjection/DependencyInjectionTransfer.java:77]: I had also assumed that the checker was going to automatically transfer any information about the right-hand side of an assignment to the left-hand side, but I wasn't seeing that behavior.
```java
      AnnotatedBindingBuilder<String> stringBinding = bind(String.class);
      stringBinding.toInstance("jdbc:mysql://localhost/pizza");
```
Before implementing `visitAssignment`, `stringBinding` _was not_ annotated with `@Bind()`

File: src/main/java/org/checkerframework/checker/dependencyinjection/DependencyInjectionAnnotatedTypeFactory.java [hunk 5]
```diff
@@ -318,48 +324,27 @@ private void handleAnnotatedWithMethod(
...
-          }
-        }
-      }
+    CFValue value = this.getStoreBefore(currentNode).getValue(JavaExpression.fromNode(receiver));
+    System.out.printf("Value: %s\n", value);
+
+    AnnotationMirror receiverAnnotation = null;
+
+    if (value != null && !value.getAnnotations().isEmpty()) {
+      receiverAnnotation = value.getAnnotations().first();
```
[the remark this answers is not in the document]
Comment (jaysonacosta) [reply src/main/java/org/checkerframework/checker/dependencyinjection/DependencyInjectionAnnotatedTypeFactory.java:342]: Got it.

File: src/main/java/org/checkerframework/checker/dependencyinjection/qual/Bind.java (hunk 4 above)
Comment (jaysonacosta) [reply src/main/java/org/checkerframework/checker/dependencyinjection/qual/Bind.java:23]: It definitely is important. I didn't know what to keep and what to leave out. I'll add the format and array size back into it.

File: src/main/java/org/checkerframework/checker/dependencyinjection/DependencyInjectionAnnotatedTypeFactory.java [hunk 6]
```diff
@@ -243,66 +239,76 @@ private void handleToMethodInvocation(
...
-              DependencyInjectionAnnotatedTypeFactory.knownBindings.remove(boundClassName);
-              // Class that is being bound to - put in knownBindings
-              // <bound,boundTo>
-              TypeMirror boundToClassName = methodArgumentNode.getType();
-
-              KnownBindingsValue knowBindingValue =
-                  new KnownBindingsValue.KnownBindingsValueBuilder(boundToClassName.toString())
-                      .setAnnotationName(annotatedNames.get(0))
-                      .build();
+  private void handleBAWAnnotation(AnnotationMirror bindAnno, Node toInstanceMethodArgumentNode) {
```
[the remark this answers is not in the document]
Comment (kelloggm) [reply src/main/java/org/checkerframework/checker/dependencyinjection/DependencyInjectionAnnotatedTypeFactory.java:256]: This hasn't been resolved: the name is still `bindAnno`.

File: src/main/java/org/checkerframework/checker/dependencyinjection/DependencyInjectionAnnotatedTypeFactory.java [hunk 7]
```diff
@@ -318,49 +324,31 @@ private void handleAnnotatedWithMethod(
...
-                TypeMirror boundToClassName = methodArgumentNode.getType();
-
-                DependencyInjectionAnnotatedTypeFactory.knownBindings.put(
-                    boundClassName,
-                    new KnownBindingsValue.KnownBindingsValueBuilder(boundToClassName.toString())
-                        .build());
-              }
-            });
-      }
+    CFValue value = this.getStoreBefore(currentNode).getValue(JavaExpression.fromNode(receiver));
```
Comment (kelloggm) [inline src/main/java/org/checkerframework/checker/dependencyinjection/DependencyInjectionAnnotatedTypeFactory.java:339]: This is computed twice: here and above at line 336.

**Review by kelloggm — CHANGES_REQUESTED** [review]

File: src/main/java/org/checkerframework/checker/dependencyinjection/DependencyInjectionAnnotatedTypeFactory.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (jaysonacosta) [reply src/main/java/org/checkerframework/checker/dependencyinjection/DependencyInjectionAnnotatedTypeFactory.java:256]: Yeah, looks like I changed the wrong one by mistake 🤦‍♂️. Fixed now.
